### DISCUSSION TITLE
DigiPol All States Roadmap

### DISCUSSION THREAD

**Discussion by KipCrossing** [voteflux/digipol#216] [Planning]
The following is a roadmap for implementing bills from all the states into digipol.

## [AusBills](https://github.com/OpenGovAus/Aus-Bills)
- Finish UI updates 
  - [https://github.com/OpenGovAus/Aus-Bills/issues/34](https://github.com/OpenGovAus/Aus-Bills/issues/34)

## [API](https://github.com/voteflux/voting-app-api)

This **SHOULD** be deployed to a new endpoint and mongoDB as the current ones are very unstable.

This **MAY** be worked on in a fresh repo

- Fetch all state data using new ausbills
- Filter out covid bills
- Create new mongo DB - leave old DB alone
- Calculate the spechash of each bill and store in static mongo DB
   - /spec
- Post ballot to blockchain 
   - [https://github.com/voteflux/voting-alpha#curl-based-can-be-run-from-cli](https://github.com/voteflux/voting-alpha#curl-based-can-be-run-from-cli)
- Add parliament key. eg: “federal”, “nsw”
- Update tag spreadsheet/machine learning 
- Add tags - topics
- inject/update dynamic mongo DB with latest bill information
  - /bill

#### optional
- Fetch results from the smart contract and serve over endpoint
  - /results
- Deploy to a new endpoint/url

## DigiPol UI
- Swap out for new API endpoints
- Fetch vote counts (results) from the blockchain
- Sorting by parliament

**Comment by KipCrossing** [comment]
Have the client fetch the vote count directly from the blockchain would be better as it more decentralized and there is no need to trust a central server for the vote count. Blockchain interactions currently make the app hang and more broadly is less efficient.

Serving the vote count over the API is more efficient but less decentralized.

Up for discussion

**Reply by connorhsm** [reply]
Unless there is a ready solution to the hanging whilst interacting with the blockchain, we should strongly consider serving the vote count over the API, with plans to revert this once a solution is found or if the pause is considered acceptable for the benefit of not relying on the single API.
I believe a decent chunk of time has already been put into searching for a solution and as far as I know there has not been one found yet.

From what I can tell, the biggest blocker to DigiPol All States is the API overhaul, would be good to catch up some time this week and discuss this with you @KipCrossing I'd be interested in working on this.

**Comment by KipCrossing** [comment]
I'm thinking that the various endpoints should be split into different projects (repos). 

### ausbills-api
- endpoint for each state with a list of bills
  - bill is
  - bill information
  - spec dict
  - spechash

### digipol-results
- sends a new proposal to the blockchain
- counts the current votes for each proposal
- serves every current proposal
  - id
  - spechash
  - spec dict
  - vote counts 

---

This way, ausbills-api can be a completely separate project and be used by other services

*we solid do ausbills-api in OpenGovAus and digipol-results in VoteFlux*
